### PR TITLE
Fix connection getting timeout while idle

### PR DESCRIPTION
**Pull request by bigmontz**
When the connection goes idle, an observer is appended. And this causes the receive timeout to start to count. This behaviour causes connections timing out while being idle in the pool.

For fixing this, the connection received methods to handle its status changing to idle and back from idle. When the connection goes idle, it should stop the timeouts and do not start any timeout until it gets busy again. 

This change also impacts the `hasOngoingRequests` method, since idle connections don't have ongoing requests.

### CODE REVIEW

File: packages/bolt-connection/src/connection/connection-channel.js [hunk 1]
```diff
@@ -393,7 +394,27 @@ export default class ChannelConnection extends Connection {
...
   /**
-   * This method still here because it's used by the {@link PooledConnectionProvider}
+   * This method is used by the {@link PooledConnectionProvider}
+   *
+   * @param {any} observer
+   */
+  _setIdle (observer) {
+    this._idle = true
+    this._ch.stopReceiveTimeout()
+    return this._protocol.queueObserverIfProtocolIsNotBroken(observer)
```
Comment (robsdedude) [inline packages/bolt-connection/src/connection/connection-channel.js]: This return value isn't being used, is it?

File: packages/bolt-connection/src/connection/connection-channel.js [hunk 2]
```diff
@@ -393,7 +394,27 @@ export default class ChannelConnection extends Connection {
...
+  }
+
+  /**
+   * This method is used by the {@link PooledConnectionProvider}
+   *
+   */
```
Comment (robsdedude) [inline packages/bolt-connection/src/connection/connection-channel.js]:
```suggestion
  /**
   * This method is used by the {@link PooledConnectionProvider}
   */
```

File: packages/bolt-connection/src/channel/browser/browser-channel.js [hunk 3]
```diff
@@ -163,17 +164,23 @@ export default class WebSocketChannel {
...
-      if (this._ws && this._ws.readyState !== WS_CLOSED) {
-        this._open = false
-        this.stopReceiveTimeout()
-        this._ws.onclose = () => resolve()
-        this._ws.close()
-      } else {
-        resolve()
-      }
-    })
+    if (this._closingPromise === null) {
```
Comment (robsdedude) [inline packages/bolt-connection/src/channel/browser/browser-channel.js]: At first glance, this change seems unrelated. Is it? 

I'm pretty sure the change is, if anything, beneficial. Yet, I'd like to understand if it was necessary or precaution and whether/how it relates to the timeout fix.

**Review by robsdedude — APPROVED** [review]
:rabbit2:

File: packages/bolt-connection/src/connection/connection-channel.js (hunk 1 above)
Comment (bigmontz) [reply packages/bolt-connection/src/connection/connection-channel.js]: It isn't.

File: packages/bolt-connection/src/channel/browser/browser-channel.js (hunk 3 above)
Comment (bigmontz) [reply packages/bolt-connection/src/channel/browser/browser-channel.js]: During the tests, I noticed problems with the connection get much time to closed in browser and then other parts of the code calling the close method also, this was causing very confusing error. This code protects for this cases. 

I can move it for other PR, if it fits better.

Comment (robsdedude) [reply packages/bolt-connection/src/channel/browser/browser-channel.js]: Sure would be cleaner to move it into a separate PR, but I don't mind sneaking in a small fix ;)

Nice catch either way :D

Comment (bigmontz) [reply packages/bolt-connection/src/channel/browser/browser-channel.js]: Moved to other PR https://github.com/neo4j/neo4j-javascript-driver/pull/1169
